Placeholder for Blog for Service CIDR (#49989)
kubernetes/website: content/en/docs/tasks/network/reconfigure-service-ip-ranges.md

---
reviewers:
- thockin
- dwinship
min-kubernetes-server-version: v1.33
title: Kubernetes Service CIDR Reconfiguration
content_type: task
---

<!-- overview -->
{{< feature-state feature_gate_name="MultiCIDRServiceAllocator" >}}

This document shares how to reconfigure the existing Service IP range assigned
to a cluster.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

{{< version-check >}}

<!-- steps -->

> ## Kubernetes Service CIDR Reconfiguration

Reviewer: This doesn't make it clear that you're talking about "changes to service CIDRs _other than_ just adding additional ones". And I'm not sure it really makes sense to have this on a separate page from the simple case anyway.

I'd break the use cases into four groups:
- adding and removing ServiceCIDRs (easy)
- single-to-dual-stack conversion preserving the primary service CIDR (a bit harder, requires an apiserver config change)
- dual-to-single conversion preserving the primary service CIDR (harder, requires fixing up existing dual-stack services)
- anything that results in changing the primary service CIDR ( :exploding_head: )

Author: +1 to the four groups, regarding having a separate page I tried to merge them,  but I had the feeling the document was too large and dense.
Should I add a subsection here and explain briefly and refer to the other for more details? @sftim some suggestion from the docs perspective?

Reviewer: (sftim is now @lmktfy)

Author: took a different angle and focused on the `Default ServiceCIDR` configuration here only, since is the one that is involved on the kube-apiserver flags

This document focuses on the dynamic reconfiguration of the Kubernetes

> Service IP address range, a capability introduced by KEP-1880. Reconfiguring

Reviewer: (Do we talk about KEPs in the end-user documentation? Does referring to KEP-1880 actually help the user here?)

Author: yeah, I should refer to the feature better

| IPv6-IPv4             | IPv4-IPv6             | High       | Swap primary IP family (IPv6 to IPv4).        |
| IPv4 only             | IPv4-IPv6             | Medium     | Introduce IPv6 as secondary IP family.        |
| IPv4 only             | IPv6-IPv4             | High       | Introduce IPv6 as primary IP family.          |
| IPv6 only             | IPv6-IPv4             | Medium     | Introduce IPv4 as secondary IP family.        |
| IPv6 only             | IPv4-IPv6             | High       | Introduce IPv4 as primary IP family.          |
| IPv4-IPv6             | IPv4 only             | High       | Remove IPv6 support.                            |
| IPv6-IPv4             | IPv6 only             | High       | Remove IPv4 support.                            |
| IPv4-IPv6             | IPv6-IPv4             | High       | Swap primary IP family (IPv4 to IPv6).        |
| IPv6-IPv4             | IPv4-IPv6             | High       | Swap primary IP family (IPv6 to IPv4).        |
| IPv4-IPv6             | IPv4-IPv6 (different CIDRs) | Medium/High | Update CIDR ranges in dual-stack.           |

> | IPv6-IPv4             | IPv6-IPv4 (different CIDRs) | Medium/High | Update CIDR ranges in dual-stack.           |

Reviewer: There's nothing particularly dual-stack about these last two cases; it's exactly the same as the single-stack case isn't it?

Author: yeah, this should be removed,

Potential issues:

* Disruptive `kubernetes.default` IP change when swaping the primary IP family.
* Incompatible existing single-stack Services.
* DNS resolution problems.
* Network plugin and external dependency updates required.
* Workload compatibility issues.
* API server log spam.
* Unpredictable behavior of network dependencies.

### The Supported Path for Reconfiguring the Service CIDR

The supported path for reconfiguring the Service CIDR involves a coordinated
update of the kube-apiserver and the cluster components, including the network plugin. Due to the tight
integration of networking components in Kubernetes, simply changing the
kube-apiserver configuration is insufficient. The network plugin must be
explicitly configured to handle the new Service IP range to ensure proper
service reachability.

### Manual Operations for Replacing the Default Service CIDR

Reconfiguring the Service CIDR necessitates manual steps performed by the
cluster operator, administrator, or the software managing the cluster lifecycle.
These typically include:

1.  **Updating** the kube-apiserver configuration: Modify the

>     `--service-cluster-ip-range` flag (and `--service-cluster-ip-ranges` for

Reviewer: ```suggestion
    `--service-cluster-ip-range` flag (or `--service-cluster-ip-ranges` for
```

Author: ```suggestion
    `--service-cluster-ip-range` flag  with the new IP range (or comma separated IP ranges for
```

Author: this was wrong, generated some content with ai and it seems it got smarter 😄

> 1.  **Updating** the kube-apiserver configuration: Modify the
>     `--service-cluster-ip-range` flag (and `--service-cluster-ip-ranges` for
>     dual-stack) with the new IP range(s).

Author: ```suggestion
1.  **Updating** the kube-apiserver configuration: Modify the
    `--service-cluster-ip-range` flag with the new IP range(s).
```

2.  **Reconfiguring** the network plugin: This is a crucial step and the
    specific procedure depends on the CNI plugin in use.
    It might involve updating configuration files, restarting agent pods, or
    applying new manifests to instruct the plugin to manage the new Service

>     CIDR.

Reviewer: There may be other network-related components in the cluster that need to be updated as well. (eg, a service mesh or multi-cluster connectivity service).

Should also clarify that there _might_ not actually be anything to do here; the network plugin might not care about the service network, or it might watch ServiceCIDR objects and update its configuration correctly automatically.

Author: yeah, I focused too much on the network plugin because of the Services/EndpointSlices relation ,  but I need to make this more generic

1.  Start the kube-apiserver with the initial `--service-cluster-ip-range`.
2.  Create initial Services that obtain IPs from this range.
3.  Introduce a new Service CIDR as a temporary target for reconfiguration.
4.  Mark the original Service CIDR for deletion (it will remain pending due to
    existing IPs and finalizers). This prevents new allocations from the old
    range.
5.  Recreate existing Services. They should now be allocated IPs from the new,
    temporary Service CIDR.
6.  Restart the kube-apiserver with the new Service CIDR(s) configured and shut
    down the old instance.
7.  Delete the `kubernetes.default` service. The new kube-apiserver will
    recreate it within the new Service CIDR.

This process highlights the need to **update** the kube-apiserver and manage
existing service IPs during reconfiguration. Critically, the network plugin
must also be configured to handle the new Service CIDR for the reconfigured
services to be functional.

### Important Considerations for Service CIDR Reconfiguration

* **Network Plugin Configuration is Key:** The successful reconfiguration
  hinges on correctly configuring your network plugin to support the new Service

>   CIDR. Consult your CNI plugin's documentation for specific instructions.

Reviewer: (again, this is a "probably", not a "definitely")

Author: removed CNI only focus for "networking components" in general
